llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td

```diff
@@ -1105,6 +1105,99 @@ def IREELinalgExt_WinogradOutputTransformOp : IREELinalgExt_Op<"winograd.output_
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Flash Attention ops
+//===----------------------------------------------------------------------===//
+
+def IREELinalgExt_FlashAttentionFwdOp : IREELinalgExt_Op<"flash_attention.fwd",
+    [DeclareOpInterfaceMethods<ReifyRankedShapedTypeOpInterface>,
+     DeclareOpInterfaceMethods<TilingInterface,
+      ["getIterationDomain",
+       "getLoopIteratorTypes",
+       "getResultTilePosition",
+       "getTiledImplementation"]>]> {
+  let summary = "Flash Attention Forward operator";
+  let description = [{
+    This is an implementation of the forward pass of the flash attention operator.
+    The operator takes 3 inputs: the query tensor (Q), the key tensor (K) and the value tensor (V).
+    All 3 inputs have dimensions BxNxd where B is the product of the batch dimension and the
+    number of heads, N is the sequence length and d is the head dimension (often N>>>d).
+    The standard attention operator computes matmul(softmax(matmul(Q, transpose(K))), V).
+    This operator computes the same quantity above but does so in a way that requires much fewer
+    HBM accesses (in the context of a CUDA GPU device). Specifically, it uses tiling and fusion
+    to compute the two matmuls and softmax in one kernel. This operator usually also performs
+    scaling, masking and dropout, but we ignore these transformations for now.
+  }];
+
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+                       Variadic<AnyShaped>:$outputs
+  );
+
+  let builders = [
+    OpBuilder<(ins "ValueRange":$inputs, "ValueRange":$outputs)>
+  ];
+
+  let results = (outs Variadic<AnyRankedTensor>:$result);
+  let hasFolder = 1;
+  let assemblyFormat = [{
+    attr-dict
+    `ins` `(` $inputs `:` type($inputs) `)`
+    `outs` `(` $outputs `:` type($outputs) `)`
+    (`->` type($result)^)?
+  }];
+
+  let extraClassDeclaration = extraLinalgExtOpClassDeclaration # [{
+    Value query() {
+      return getInputOperand(0)->get();
+    }
+    Value key() {
+      return getInputOperand(1)->get();
+    }
+    Value value() {
+      return getInputOperand(2)->get();
+    }
```

Comment:
Can't the op just have:

```
let arguments = (ins AnyShaped:$query, AnyShaped:$key, AnyShaped:$value, ...)`
```
?

Reply:
See my suggestion above. The interface cannot require ops to have ODS arguments with specific names or kinds. FWIW, it's better to explicitly implement interface methods instead of rely on implicitly generated methods to fit what the interface requires.

Reply:
Looking into this more, it seems like we can't have
`let arguments = (ins AnyShaped:$query, AnyShaped:$key, AnyShaped:$value, ...)`
because of the `AttrSizedOperandSegments` trait that every LinalgExtOp needs to satisfy
https://github.com/iree-org/iree/blob/52c5d1422b2769149acea74202710eb5e9e887dc/llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td#L48. Since the variadic is still required, I don't think it makes sense to separate out into query, key and value because each one of those will be required to be variadic. I propose we stick with the current formulation until AttrSizedOperandSegments is removed from LinalgExtOps. What do you think?

Reply:
Yeah... we should drop it from the LinalgExt base op and move it into individual ops.... Just one of the things that fell through the cracks...

Reply:
Could we create a bug to track these cleanups? I don't want to block this PR on legacy refactorings.